crates/bevy_ecs/src/entity/map_entities.rs

```diff
@@ -27,19 +31,62 @@ use bevy_utils::EntityHashMap;
 /// }
 ///
 /// impl MapEntities for Spring {
-///     fn map_entities(&mut self, entity_mapper: &mut EntityMapper) {
-///         self.a = entity_mapper.get_or_reserve(self.a);
-///         self.b = entity_mapper.get_or_reserve(self.b);
+///     fn map_entities<M: Mapper>(&mut self, entity_mapper: &mut M) {
+///         self.a.map_entities(entity_mapper);
+///         self.b.map_entities(entity_mapper)
 ///     }
 /// }
 /// ```
 ///
 pub trait MapEntities {
     /// Updates all [`Entity`] references stored inside using `entity_mapper`.
-    ///
-    /// Implementors should look up any and all [`Entity`] values stored within and
-    /// update them to the mapped values via `entity_mapper`.
-    fn map_entities(&mut self, entity_mapper: &mut EntityMapper);
+    fn map_entities<M: Mapper>(&mut self, entity_mapper: &mut M);
+}
+
+/// This traits defines a type that knows how to map [`Entity`] references.
+///
+/// Two implementations are provided:
+/// - `SimpleEntityMapper`: tries to map the [`Entity`] reference, but if it can't, it returns the same [`Entity`] reference.
+/// - `EntityMapper`: tries to map the [`Entity`] reference, but if it can't, it allocates a new [`Entity`] reference.
+pub trait Mapper {
+    /// Map an entity to another entity
+    fn map(&mut self, entity: Entity) -> Entity;
+}
+
+/// Similar to `EntityMapper`, but does not allocate new [`Entity`] references in case we couldn't map the entity.
+pub struct SimpleEntityMapper<'m> {
+    map: &'m EntityHashMap<Entity, Entity>,
+}
+
+impl Mapper for SimpleEntityMapper<'_> {
+    /// Map the entity to another entity, or return the same entity if we couldn't map it.
+    fn map(&mut self, entity: Entity) -> Entity {
+        self.get(entity).unwrap_or(entity)
+    }
+}
+
+impl<'m> SimpleEntityMapper<'m> {
+    /// Creates a new `SimpleEntityMapper` from an [`EntityHashMap<Entity, Entity>`].
+    pub fn new(map: &'m EntityHashMap<Entity, Entity>) -> Self {
+        Self { map }
+    }
+
+    /// Returns the corresponding mapped entity or None if it is absent.
+    pub fn get(&self, entity: Entity) -> Option<Entity> {
+        self.map.get(&entity).copied()
+    }
+
+    /// Gets a reference to the underlying [`EntityHashMap<Entity, Entity>`].
+    pub fn get_map(&'m self) -> &'m EntityHashMap<Entity, Entity> {
+        self.map
+    }
+}
+
+impl Mapper for EntityMapper<'_> {
+    /// Returns the corresponding mapped entity or reserves a new dead entity ID if it is absent.
+    fn map(&mut self, entity: Entity) -> Entity {
+        self.get_or_reserve(entity)
```

Comment:
I would inline `get_or_reserve` here. Otherwise the struct will have two identical methods with different names.

Reply:
You mean add `#[inline]` tag on the `get_or_reserve` code?
I think duplicating the code directly could be errorprone

Reply:
I meant to move the implementation from `get_or_reserve` here.

Reply:
I'm not sure i'd want a breaking change like removing a public function; I think it's fine if the `Mapper` implementation re-uses an internal function from `EntityMapper`

Reply:
Mark the redundant method as deprecated please, then we can remove it on the next cycle with minimal pain.

Reply:
Copied the code to `map` and marked `get_or_reserve` as deprecated